src/threading.c

```diff
@@ -848,6 +848,138 @@ JL_DLLEXPORT int jl_alignment(size_t sz)
     return jl_gc_alignment(sz);
 }
 
+// Heartbeat mechanism for Julia's task scheduler
+// ---
+// Start a thread that does not participate in running Julia's tasks. This
+// thread simply sleeps until the heartbeat mechanism is enabled. When
+// enabled, the heartbeat thread enters a loop in which it waits for the
+// specified heartbeat interval. If, within that interval, `jl_heartbeat()`
+// is *not* called at least once, then the thread calls
+// `jl_print_task_backtraces(0)`.
+
+#ifdef JL_HEARTBEAT_THREAD
+
+#include <time.h>
+
+volatile int heartbeat_enabled;
+uv_sem_t heartbeat_sem;
+int heartbeat_interval_s, report_interval_s, last_report_s;
+_Atomic(int) heartbeats;
+
+JL_DLLEXPORT void jl_print_task_backtraces(int show_done) JL_NOTSAFEPOINT;
+void jl_heartbeat_threadfun(void *arg);
+
+// start the heartbeat thread with heartbeats disabled
+void jl_init_heartbeat(void)
+{
+    uv_thread_t uvtid;
+    heartbeat_enabled = 0;
+    uv_sem_init(&heartbeat_sem, 0);
+    uv_thread_create(&uvtid, jl_heartbeat_threadfun, NULL);
+    uv_thread_detach(&uvtid);
+}
+
+// enable/disable heartbeats
+JL_DLLEXPORT int jl_heartbeat_enable(int heartbeat_s, int report_s)
+{
+    if (heartbeat_s <= 0) {
+        heartbeat_enabled = 0;
+        last_report_s = 0;
+    }
+    else {
+        jl_atomic_store_relaxed(&heartbeats, 0);
+        heartbeat_interval_s = heartbeat_s;
+        report_interval_s = report_s;
+        heartbeat_enabled = 1;
+        uv_sem_post(&heartbeat_sem);
+    }
+    return 0;
+}
+
+// heartbeat
+JL_DLLEXPORT void jl_heartbeat(void)
+{
+    jl_atomic_fetch_add(&heartbeats, 1);
+}
+
+// sleep the thread for the specified interval
+void sleep_for(int secs, int nsecs)
+{
+    struct timespec rqtp, rmtp;
+    rqtp.tv_sec = secs;
+    rqtp.tv_nsec = nsecs;
+    rmtp.tv_sec = 0;
+    rmtp.tv_nsec = 0;
+    for (; ;) {
+        if (nanosleep(&rqtp, &rmtp) == 0)
+            return;
+        rqtp = rmtp;
+    }
+}
+
+uint8_t check_heartbeats(uint8_t gc_state)
+{
+    int hb = jl_atomic_exchange(&heartbeats, 0);
+
+    if (hb <= 0) {
+        if ((jl_hrtime() / 1e9) - last_report_s >= report_interval_s) {
+            jl_task_t *ct = jl_current_task;
+            jl_ptls_t ptls = ct->ptls;
+            jl_gc_safe_leave(ptls, gc_state);
+            jl_print_task_backtraces(0);
+            gc_state = jl_gc_safe_enter(ptls);
+            last_report_s = jl_hrtime() / 1e9;
+        }
```

Comment:
IMO it's nice to have some kind of exponential backoff on the reporting interval for expensive reports like this. Something like `report_interval_s *= 1.5` or something.

Up to you though. If you set it large enough it's probably fine, but I think that reporting the task backtraces can be pretty slow for us in raicode, since stack unwinding + symbolizing is really slow in our huge binary, and I'm worried that this brownout reporting can itself cause some brownouts? 😰 So I would probably suggest backing off the reporting interval with each report?

And then maybe shortening the interval again if we get one and it's been like 100x the reporting interval or something?
I guess this can be kind of finicky, but it's the kind of trick I've seen @tveldhui use to reduce the impact of expensive logs like this.

Reply:
Added a 1.5X backoff to the reporting interval up to a maximum of 50X the requested reporting interval.

Reply:
Changed this to a 2X backoff in reporting time.